components/raftstore-v2/src/operation/query/capture.rs

```diff
@@ -0,0 +1,124 @@
+// Copyright 2023 TiKV Project Authors. Licensed under Apache-2.0.
+
+use std::sync::Arc;
+
+use engine_traits::{KvEngine, RaftEngine};
+use fail::fail_point;
+use kvproto::raft_cmdpb::RaftCmdResponse;
+use raftstore::{
+    coprocessor::ObserveHandle,
+    store::{
+        cmd_resp,
+        fsm::{
+            apply::{notify_stale_req_with_msg, ObserverType},
+            new_read_index_request, ChangeObserver,
+        },
+        msg::ErrorCallback,
+        util::compare_region_epoch,
+        RegionSnapshot,
+    },
+};
+
+use crate::{
+    fsm::{ApplyResReporter, PeerFsmDelegate},
+    raft::Apply,
+    router::{message::CaptureChange, ApplyTask, QueryResChannel, QueryResult},
+};
+
+impl<'a, EK: KvEngine, ER: RaftEngine, T: raftstore::store::Transport>
+    PeerFsmDelegate<'a, EK, ER, T>
+{
+    pub fn on_leader_callback(&mut self, ch: QueryResChannel) {
+        let peer = self.fsm.peer();
+        let msg = new_read_index_request(
+            peer.region_id(),
+            peer.region().get_region_epoch().clone(),
+            peer.peer().clone(),
+        );
+        self.on_query(msg, ch);
+    }
+
+    pub fn on_capture_change(&mut self, capture_change: CaptureChange) {
+        fail_point!("raft_on_capture_change");
+
+        // TODO: Allow to capture change even is in flashback state.
+        // TODO: add a test case for this kind of situation.
+
+        let apply_router = self.fsm.peer().apply_scheduler().unwrap().clone();
+        let (ch, _) = QueryResChannel::with_callback(Box::new(move |res| {
+            if let QueryResult::Response(resp) = res && resp.get_header().has_error() {
+                // Return error
+                capture_change.snap_cb.report_error(resp.clone());
+                return;
+            }
+            apply_router.send(ApplyTask::CaptureApply(capture_change))
```

Comment:
A (maybe dummy) question: if we asynchronously change the capture state, is it possible to lose data between `[read_index_idx, change_capture_state_idx)`? Or `apply_router` has a synchronous nature?

Reply:
The read index here is fow 1) ensure it's a leader and 2) the its data is ready to read. The data between `[read_index_idx, change_capture_state_idx)` is included in snapshot which is taken by `Apply`.

Reply:
I see, we get the snapshot when handling the `CaptureChange` instead of after `ReadIndex`.